### ISSUE TITLE
problem with Dogwater using Nes with JWT-authentication

### ISSUE THREAD

**Issue by finedays** [devinivy/dogwater#53]
Hello guys, 

I have some troubles using NES with jwt authentication using Waterline/Dogwater.

here is my server.js:
```
'use strict'
const Hapi = require('hapi')
const Hoek = require('hoek')
const hapiAuthJWT = require('hapi-auth-jwt2')
const Dogwater = require('dogwater')
const sailsmysql = require('sails-mysql')
const Nes = require('nes')

// load models

const config = require('./configs/test_credentials')
const router = require('./routes/router').router
const validate = require('./lib/validate').validate

const server = new Hapi.Server()
server.connection({ port: 8000, routes: { cors: true } })

server.register(require('inert'), (err) => {
  Hoek.assert(!err, err)
})

server.register({
  register: Dogwater,
  options: {
    adapters: {
      dndDisk: sailsmysql
    },
    connections: {
      dndDB: .....................
    }
  }
}, (err) => {
  Hoek.assert(!err, err)
})

server.register(hapiAuthJWT, (err) => {
  Hoek.assert(!err, err)
})

server.auth.strategy('jwt', 'jwt',
  { key: config.jwtsecret,
    validateFunc: validate,
    verifyOptions: { ignoreExpiration: true },
    headerKey: 'authentication-token'
  })

server.register(Nes, (err) => {
  Hoek.assert(!err, err)
})

server.auth.default('jwt')

server.dogwater([ CreditCard, Passenger, FavoriteLocation, Trip, Driver, TripRoute, PaymentLog, Promotion, DriverRating, PassengerRating, Administrator, Country, CountryPrice, LatestDriverLocation, Language ])

server.route(router)

server.start((err) => {
  Hoek.assert(!err, err)
  console.log('Server started at:', server.info.uri)
})
```

My route for the socket is:
```
{
    method: 'GET',
    path: '/ws',
    config: {
      id: 'hello',
      auth: 'jwt',
      handler: function (request, reply) {

                return reply('world!');
            }
    }
  }
```

validate function is: 
```
function validate (decoded, request, callback) {
  const passenger = request.collections().passenger
  passenger.findOne(decoded.id)
    .then((passenger) => {
      if ((decoded.username) === (passenger.username)) {
        return callback(null, true)
      } else {
        return callback(null, false)
      }
    })
}
```

and client is: 
```
var Nes = require('nes')
var client = new Nes.Client('ws://localhost:8000')

client.connect({auth: {headers: {'Authentication-Token': 'token'}}}, (err) => {
  if (err) return console.log(err)
  client.request('hello', (err, payload) => {
    if (err) return console.log('err is ' + err)
    console.log(payload)
  })
})
```

when I run the client, using valid token (I am testing it with other routes) it gives the following error:

> { Error: An internal server error occurred
    at new NesError (/Users/vgenev/Projects/dnd-server/node_modules/nes/lib/client.js:77:19)
    at parse (/Users/vgenev/Projects/dnd-server/node_modules/nes/lib/client.js:616:25)
    at parse (/Users/vgenev/Projects/dnd-server/node_modules/nes/lib/client.js:48:16)
    at Client._onMessage (/Users/vgenev/Projects/dnd-server/node_modules/nes/lib/client.js:603:9)
    at WebSocket.ws.onmessage (/Users/vgenev/Projects/dnd-server/node_modules/nes/lib/client.js:261:25)
    at WebSocket.onMessage (/Users/vgenev/Projects/dnd-server/node_modules/ws/lib/WebSocket.js:442:14)
    at emitTwo (events.js:106:13)
    at WebSocket.emit (events.js:191:7)
    at Receiver.ontext (/Users/vgenev/Projects/dnd-server/node_modules/ws/lib/WebSocket.js:841:10)
    at /Users/vgenev/Projects/dnd-server/node_modules/ws/lib/Receiver.js:536:18
  type: 'server',
  statusCode: 401,
  data:
   { error: 'Unauthorized',
     message: 'An internal server error occurred' },
  headers: undefined,
  path: undefined }

and the server error is:

> Debug: internal, implementation, error 
    TypeError: Cannot read property 'findOne' of undefined
    at Object.validate [as validateFunc] (/Users/vgenev/Projects/dnd-server/lib/validate.js:4:12)
    at /Users/vgenev/Projects/dnd-server/node_modules/hapi-auth-jwt2/lib/index.js:87:23
    at /Users/vgenev/Projects/dnd-server/node_modules/jsonwebtoken/verify.js:27:18
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
    at process._tickDomainCallback (internal/process/next_tick.js:122:9)

What I have seen is that actually the passenger is not created by: `  const passenger = request.collections().passenger` in the validate function, although the token is correctly decoded.

After I posted the question in Nes repo, they adviced me to use `request.collections(true).passenger` and it fixed the problem. But is it a good practice and how can be avoided?

**Comment by devinivy** [comment]
Hi!  It's me again :)
Dogwater is structured so that each model is only available to the plugin that registered it.  If you registered models to the root server (outside a plugin), then those models will only be available outside of a plugin.  As I hinted earlier, passing `true` to `request.collections()` sidesteps this sandboxing constraint.  I checked and played with the dogwater tests to ensure this functionality is working as intended.  So, it leads me to the question– is the route registered inside of a plugin?

It turns out that by default, nes creates a route specifically for authenticating the socket connection (see auth section with info on "direct" authentication [here](https://github.com/hapijs/nes/blob/master/API.md#registration)).  Because that route is created by the nes plugin (which naturally doesn't register any dogwater models), there is no access to dogwater models inside your `validateFunc` (where `request` is a request to the route created by nes)!

In short, this is a good reason to use `request.collections(true)` to get access to all waterline models, ignoring plugin boundaries.  A route created by nes will otherwise not be able to access any models.  It's a pretty rare case!  This is exactly why I added that flag, though.

I think your instinct is good– in general models will only be used by the plugins that register them using `server.dogwater()`, so passing that flag to `request.collections()` is typically not necessary.  If you find yourself having to use that flag often, there could be some issues with your plugin architecture creating too many inter-plugin dependencies.

**Comment by finedays** [comment]
Thanks a lot :) I will let it sink and work more with both and see should I redesign the plugin architecture.
